File: ios-mobile-app/Vuga/Assets/Managers/ConnectionMonitor.swift
import Foundation
import Network
import SwiftUI
import Combine

enum ConnectionQuality {
    case excellent
    case good
    case fair
    case poor
    case offline
    
    var displayText: String {
        switch self {
        case .excellent:
            return "Excellent Connection"
        case .good:
            return "Good Connection"
        case .fair:
            return "Slow Connection"
        case .poor:
            return "Very Slow Connection"
        case .offline:
            return "No Internet Connection"
        }
    }
    
    var color: Color {
        switch self {
        case .excellent:
            return .green
        case .good:
            return .green
        case .fair:
            return .orange
        case .poor:
            return .red
        case .offline:
            return .red
        }
    }
    
    var iconName: String {
        switch self {
        case .excellent:
            return "wifi"
        case .good:
            return "wifi"
        case .fair:
            return "wifi.exclamationmark"
        case .poor:
            return "wifi.exclamationmark"
        case .offline:
            return "wifi.slash"
        }
    }
}

class ConnectionMonitor: ObservableObject {
    static let shared = ConnectionMonitor()
    
    private let monitor = NWPathMonitor()
    private let queue = DispatchQueue(label: "NetworkMonitor")
    private var cancellables = Set<AnyCancellable>()
    
    @Published var isConnected: Bool = true
    @Published var connectionType: NWInterface.InterfaceType?
    @Published var connectionQuality: ConnectionQuality = .good
    @Published var showConnectionAlert: Bool = false
    @Published var latency: TimeInterval = 0
    @Published var downloadSpeed: Double = 0 // Mbps
    
    // Track API response times for quality assessment
    private var recentResponseTimes: [TimeInterval] = []
    private let maxResponseTimeSamples = 10
    
    // Thresholds for connection quality (in seconds)
    private let excellentThreshold: TimeInterval = 0.5
    private let goodThreshold: TimeInterval = 1.0
    private let fairThreshold: TimeInterval = 2.0
    private let poorThreshold: TimeInterval = 4.0
    
    private init() {
        startMonitoring()
    }
    
    private func startMonitoring() {
        monitor.pathUpdateHandler = { [weak self] path in
            DispatchQueue.main.async {
                self?.updateConnectionStatus(path)
            }
        }
        
        monitor.start(queue: queue)
        
        // Periodically check connection quality
        Timer.publish(every: 30, on: .main, in: .common)
            .autoconnect()
            .sink { [weak self] _ in
                self?.checkConnectionQuality()
            }
            .store(in: &cancellables)
    }
    
    private func updateConnectionStatus(_ path: NWPath) {
        let wasConnected = isConnected
        isConnected = path.status == .satisfied
        
        if path.usesInterfaceType(.wifi) {
            connectionType = .wifi
        } else if path.usesInterfaceType(.cellular) {
            connectionType = .cellular
        } else {
            connectionType = nil
        }
        
        // Show alert when connection status changes
        if wasConnected && !isConnected {
            connectionQuality = .offline
            showConnectionAlert = true
            sendConnectionNotification(connected: false)
        } else if !wasConnected && isConnected {
            showConnectionAlert = false
            sendConnectionNotification(connected: true)
            checkConnectionQuality()
        }
        
        // Update quality based on connection type
        if !isConnected {
            connectionQuality = .offline
        }
    }
    
    func recordResponseTime(_ time: TimeInterval) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            
            self.recentResponseTimes.append(time)
            if self.recentResponseTimes.count > self.maxResponseTimeSamples {
                self.recentResponseTimes.removeFirst()
            }
            
            self.updateConnectionQuality()
        }
    }
    
    private func updateConnectionQuality() {
        guard isConnected else {
            connectionQuality = .offline
            return
        }
        
        guard !recentResponseTimes.isEmpty else {
            // Default to good if no data yet
            connectionQuality = .good
            return
        }
        
        let averageResponseTime = recentResponseTimes.reduce(0, +) / Double(recentResponseTimes.count)
        latency = averageResponseTime
        
        // Determine quality based on average response time
        let previousQuality = connectionQuality
        
        switch averageResponseTime {
        case 0..<excellentThreshold:
            connectionQuality = .excellent
        case excellentThreshold..<goodThreshold:
            connectionQuality = .good
        case goodThreshold..<fairThreshold:
            connectionQuality = .fair
        case fairThreshold..<poorThreshold:
            connectionQuality = .poor
        default:
            connectionQuality = .poor
        }
        
        // Show alert if connection becomes poor
        if previousQuality != connectionQuality &&
           (connectionQuality == .poor || connectionQuality == .fair) {
            showConnectionAlert = true
            
            // Auto-hide after 3 seconds for non-critical alerts
            DispatchQueue.main.asyncAfter(deadline: .now() + 3) { [weak self] in
                if self?.connectionQuality != .offline {
                    self?.showConnectionAlert = false
                }
            }
        }
    }
    
    func checkConnectionQuality() {
        guard isConnected else { return }
        
        // Ping a reliable server to test latency
        testLatency()
    }
    
    private func testLatency() {
        let url = URL(string: "https://www.google.com")!
        let startTime = Date()
        
        var request = URLRequest(url: url)
        request.httpMethod = "HEAD"
        request.timeoutInterval = 5
        
        URLSession.shared.dataTask(with: request) { [weak self] _, response, error in
            let responseTime = Date().timeIntervalSince(startTime)
            
            if error == nil && response != nil {
                self?.recordResponseTime(responseTime)
            } else if error != nil {
                // Connection is very slow or problematic
                self?.recordResponseTime(5.0)
            }
        }.resume()
    }
    
    private func sendConnectionNotification(connected: Bool) {
        let notificationName = connected ? 
            NSNotification.Name("NetworkConnected") : 
            NSNotification.Name("NetworkDisconnected")
        
        NotificationCenter.default.post(name: notificationName, object: nil)
    }
    
    func startSpeedTest(completion: @escaping (Double) -> Void) {
        // Simple speed test using a small file download
        let testURL = URL(string: "https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png")!
        let startTime = Date()
        
        URLSession.shared.dataTask(with: testURL) { data, _, error in
            guard let data = data, error == nil else {
                completion(0)
                return
            }
            
            let elapsedTime = Date().timeIntervalSince(startTime)
            let bytes = Double(data.count)
            let bits = bytes * 8
            let megabits = bits / 1_000_000
            let speed = megabits / elapsedTime
            
            DispatchQueue.main.async {
                self.downloadSpeed = speed
                completion(speed)
            }
        }.resume()
    }
}

File: android-mobile-app/app/src/main/java/com/retry/vuga/utils/ConnectionMonitor.java
package com.retry.vuga.utils;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkRequest;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import androidx.annotation.NonNull;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class ConnectionMonitor {
    private static final String TAG = "ConnectionMonitor";
    private static ConnectionMonitor instance;
    
    private Context context;
    private ConnectivityManager connectivityManager;
    private NetworkCallbackImpl networkCallback;
    private Handler mainHandler;
    private ExecutorService executor;
    private OkHttpClient httpClient;
    
    // Connection state
    private boolean isConnected = true;
    private ConnectionType connectionType = ConnectionType.WIFI;
    private ConnectionQuality connectionQuality = ConnectionQuality.GOOD;
    private boolean showConnectionAlert = false;
    private long latency = 0;
    private double downloadSpeed = 0.0;
    
    // Response time tracking
    private List<Long> recentResponseTimes = new ArrayList<>();
    private static final int MAX_RESPONSE_TIME_SAMPLES = 10;
    
    // Quality thresholds (in milliseconds)
    private static final long EXCELLENT_THRESHOLD = 500;
    private static final long GOOD_THRESHOLD = 1000;
    private static final long FAIR_THRESHOLD = 2000;
    private static final long POOR_THRESHOLD = 4000;
    
    // Listeners
    private List<ConnectionListener> listeners = new ArrayList<>();
    
    public enum ConnectionType {
        WIFI, CELLULAR, NONE
    }
    
    public enum ConnectionQuality {
        EXCELLENT("Excellent Connection"),
        GOOD("Good Connection"),
        FAIR("Slow Connection"),
        POOR("Very Slow Connection"),
        OFFLINE("No Internet Connection");
        
        private final String displayText;
        
        ConnectionQuality(String displayText) {
            this.displayText = displayText;
        }
        
        public String getDisplayText() {
            return displayText;
        }
    }
    
    public interface ConnectionListener {
        void onConnectionChanged(boolean isConnected, ConnectionType type, ConnectionQuality quality);
        void onConnectionAlert(boolean show, String message);
    }
    
    private ConnectionMonitor(Context context) {
        this.context = context.getApplicationContext();
        this.connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        this.mainHandler = new Handler(Looper.getMainLooper());
        this.executor = Executors.newSingleThreadExecutor();
        this.httpClient = new OkHttpClient();
        this.networkCallback = new NetworkCallbackImpl();
        
        startMonitoring();
    }
    
    public static synchronized ConnectionMonitor getInstance(Context context) {
        if (instance == null) {
            instance = new ConnectionMonitor(context);
        }
        return instance;
    }
    
    public static ConnectionMonitor getInstance() {
        if (instance == null) {
            throw new IllegalStateException("ConnectionMonitor not initialized. Call getInstance(Context) first.");
        }
        return instance;
    }
    
    private void startMonitoring() {
        NetworkRequest.Builder builder = new NetworkRequest.Builder()
                .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET);
        
        connectivityManager.registerNetworkCallback(builder.build(), networkCallback);
        
        // Check initial state
        updateConnectionStatus();
        
        // Start periodic quality checks
        startPeriodicQualityCheck();
    }
    
    private void startPeriodicQualityCheck() {
        mainHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                if (isConnected) {
                    checkConnectionQuality();
                }
                mainHandler.postDelayed(this, 30000); // Check every 30 seconds
            }
        }, 30000);
    }
    
    private void updateConnectionStatus() {
        Network activeNetwork = connectivityManager.getActiveNetwork();
        boolean wasConnected = isConnected;
        
        if (activeNetwork != null) {
            NetworkCapabilities capabilities = connectivityManager.getNetworkCapabilities(activeNetwork);
            if (capabilities != null) {
                isConnected = capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET);
                
                if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
                    connectionType = ConnectionType.WIFI;
                } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) {
                    connectionType = ConnectionType.CELLULAR;
                } else {
                    connectionType = ConnectionType.NONE;
                }
            } else {
                isConnected = false;
                connectionType = ConnectionType.NONE;
            }
        } else {
            isConnected = false;
            connectionType = ConnectionType.NONE;
        }
        
        // Update quality based on connection status
        if (!isConnected) {
            connectionQuality = ConnectionQuality.OFFLINE;
            showConnectionAlert = true;
            notifyConnectionAlert(true, "No internet connection");
        } else if (!wasConnected && isConnected) {
            showConnectionAlert = false;
            notifyConnectionAlert(false, "");
            checkConnectionQuality();
        }
        
        // Notify listeners
        notifyConnectionChanged();
        
        Log.d(TAG, "Connection status - Connected: " + isConnected + ", Type: " + connectionType + ", Quality: " + connectionQuality);
    }
    
    public void recordResponseTime(long responseTime) {
        mainHandler.post(() -> {
            recentResponseTimes.add(responseTime);
            if (recentResponseTimes.size() > MAX_RESPONSE_TIME_SAMPLES) {
                recentResponseTimes.remove(0);
            }
            updateConnectionQuality();
        });
    }
    
    private void updateConnectionQuality() {
        if (!isConnected) {
            connectionQuality = ConnectionQuality.OFFLINE;
            return;
        }
        
        if (recentResponseTimes.isEmpty()) {
            connectionQuality = ConnectionQuality.GOOD;
            return;
        }
        
        // Calculate average response time
        long sum = 0;
        for (Long time : recentResponseTimes) {
            sum += time;
        }
        long averageResponseTime = sum / recentResponseTimes.size();
        latency = averageResponseTime;
        
        ConnectionQuality previousQuality = connectionQuality;
        
        // Determine quality based on average response time
        if (averageResponseTime < EXCELLENT_THRESHOLD) {
            connectionQuality = ConnectionQuality.EXCELLENT;
        } else if (averageResponseTime < GOOD_THRESHOLD) {
            connectionQuality = ConnectionQuality.GOOD;
        } else if (averageResponseTime < FAIR_THRESHOLD) {
            connectionQuality = ConnectionQuality.FAIR;
        } else if (averageResponseTime < POOR_THRESHOLD) {
            connectionQuality = ConnectionQuality.POOR;
        } else {
            connectionQuality = ConnectionQuality.POOR;
        }
        
        // Show alert if connection becomes poor
        if (previousQuality != connectionQuality && 
            (connectionQuality == ConnectionQuality.POOR || connectionQuality == ConnectionQuality.FAIR)) {
            showConnectionAlert = true;
            notifyConnectionAlert(true, connectionQuality.getDisplayText());
            
            // Auto-hide after 3 seconds for non-critical alerts
            mainHandler.postDelayed(() -> {
                if (connectionQuality != ConnectionQuality.OFFLINE) {
                    showConnectionAlert = false;
                    notifyConnectionAlert(false, "");
                }
            }, 3000);
        }
        
        notifyConnectionChanged();
    }
    
    private void checkConnectionQuality() {
        if (!isConnected) return;
        
        executor.execute(this::testLatency);
    }
    
    private void testLatency() {
        Request request = new Request.Builder()
                .url("https://www.google.com")
                .head()
                .build();
        
        long startTime = System.currentTimeMillis();
        
        httpClient.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                // Connection is very slow or problematic
                recordResponseTime(5000);
            }
            
            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) {
                long responseTime = System.currentTimeMillis() - startTime;
                recordResponseTime(responseTime);
                response.close();
            }
        });
    }
    
    public void startSpeedTest(SpeedTestCallback callback) {
        if (!isConnected) {
            callback.onResult(0.0);
            return;
        }
        
        Request request = new Request.Builder()
                .url("https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png")
                .build();
        
        long startTime = System.currentTimeMillis();
        
        httpClient.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                mainHandler.post(() -> {
                    downloadSpeed = 0.0;
                    callback.onResult(0.0);
                });
            }
            
            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) {
                try {
                    if (response.body() != null) {
                        byte[] bytes = response.body().bytes();
                        long elapsedTime = System.currentTimeMillis() - startTime;
                        
                        double bytesDownloaded = bytes.length;
                        double bits = bytesDownloaded * 8;
                        double megabits = bits / 1_000_000;
                        double speed = megabits / (elapsedTime / 1000.0);
                        
                        mainHandler.post(() -> {
                            downloadSpeed = speed;
                            callback.onResult(speed);
                        });
                    }
                } catch (IOException e) {
                    mainHandler.post(() -> {
                        downloadSpeed = 0.0;
                        callback.onResult(0.0);
                    });
                } finally {
                    response.close();
                }
            }
        });
    }
    
    public void addListener(ConnectionListener listener) {
        if (!listeners.contains(listener)) {
            listeners.add(listener);
        }
    }
    
    public void removeListener(ConnectionListener listener) {
        listeners.remove(listener);
    }
    
    private void notifyConnectionChanged() {
        for (ConnectionListener listener : listeners) {
            listener.onConnectionChanged(isConnected, connectionType, connectionQuality);
        }
    }
    
    private void notifyConnectionAlert(boolean show, String message) {
        for (ConnectionListener listener : listeners) {
            listener.onConnectionAlert(show, message);
        }
    }
    
    // Getters
    public boolean isConnected() {
        return isConnected;
    }
    
    public ConnectionType getConnectionType() {
        return connectionType;
    }
    
    public ConnectionQuality getConnectionQuality() {
        return connectionQuality;
    }
    
    public boolean shouldShowConnectionAlert() {
        return showConnectionAlert;
    }
    
    public long getLatency() {
        return latency;
    }
    
    public double getDownloadSpeed() {
        return downloadSpeed;
    }
    
    public void stopMonitoring() {
        if (connectivityManager != null && networkCallback != null) {
            connectivityManager.unregisterNetworkCallback(networkCallback);
        }
        if (executor != null) {
            executor.shutdown();
        }
    }
    
    public interface SpeedTestCallback {
        void onResult(double speedMbps);
    }
    
    private class NetworkCallbackImpl extends ConnectivityManager.NetworkCallback {
        @Override
        public void onAvailable(@NonNull Network network) {
            Log.d(TAG, "Network available: " + network);
            mainHandler.post(() -> updateConnectionStatus());
        }
        
        @Override
        public void onLost(@NonNull Network network) {
            Log.d(TAG, "Network lost: " + network);
            mainHandler.post(() -> updateConnectionStatus());
        }
        
        @Override
        public void onCapabilitiesChanged(@NonNull Network network, @NonNull NetworkCapabilities networkCapabilities) {
            Log.d(TAG, "Network capabilities changed: " + network);
            mainHandler.post(() -> updateConnectionStatus());
        }
    }
}